### PR TITLE
 Code to send pushover notifications (ikolubr - Mar 31st, 2018)

### PR DESCRIPTION
**Pull request by ikolubr**
Code to send pushover notifications (ikolubr - Mar 31st, 2018)

### CODE REVIEW

File: p2pool/work.py [hunk 1]
```diff
@@ -418,6 +420,18 @@ def got_response(header, user, coinbase_nonce):
...
+                        # Code to send pushover notifications when a block is found by a miner (ikolubr - Mar 31st, 2018)
+                        if self.node.net.USE_PUSHOVER_BLOCK:
+                            conn = httplib.HTTPSConnection("api.pushover.net:443")
+                            conn.request("POST", "/1/messages.json",
+                                urllib.urlencode({
+                                    "token": self.node.net.PUSHOVER_APP_TOKEN,
+                                    "user": self.node.net.PUSHOVER_USER_KEY,
+                                    "message": 'FOUND BLOCK! %s%064x' % (self.node.net.PARENT.BLOCK_EXPLORER_URL_PREFIX, header_hash),
+                                }), { "Content-type": "application/x-www-form-urlencoded" })
+                            conn.getresponse()
```
Comment (jtoomim) [inline p2pool/work.py]: If api.pushover.net:443 isn't responding or is slow, what happens? Does p2pool hang until the push times out?

It seems that this code is running in the wrong place. Push notifications should be sent *after* all latency-critical tasks have been completed. It looks like you're sending out the push notifications before sending out the blocks or shares to peers, which means that block propagation (a revenue-critical and latency-sensitive process) is delayed until after push notification (an optional cosmetic feature) is completed.

**Comment by ikolubr** [comment]
You are completely right. I will do some coding and get back to you! Thank you.

**Comment by jtoomim** [comment]
An example of how to do command-line options can be found in this commit:

https://github.com/jtoomim/p2pool/commit/9692d6e8f9980b057ae67e8970353be3411fe0fe#diff-a9b0b3b92c88cf385af4c156ba6fb8deR459

main.py:459 and :561

**Comment by ikolubr** [comment]
man I don't know if this is the right place to ask this... if you prefer a more private way let me know...
I'm trying to understand p2pool as a whole so I can try to help you with developement.
Another day I was talking with a few friends and we kinda don't understand how p2pool generates a block and how the shares interfere with that or not.

Like, who finds a block? Is it p2pool or a miner? Does p2pool kinda merges all the shares to create a valid block or the only purpose of the shares are for the payout calculations? If that is true, is each miner kinda mining solo and when it finds a block p2pool split that between all the miners that were mining "solo" as well? Also if that is true, why we care about shares diff? Isn't the pseudo share what we should be caring about?! Isn't the pseudo share what works towards to find a block?

Sorry for all the doubts and questions... I'm really just trying to understand as a whole so I can put my mind to work and focus on it. Knowing how it works helps me a lot to picture new code.

Thank you so much,

**Comment by jtoomim** [comment]
There is no "towards". There is no progress. There is only the hash value. If the hash is below the share target, then it can be a valid share. If the hash is below the block target, then it can be both a valid share and a valid block. (Share target > block target). Pseudoshares are only for statistics collection for graphs and such and have no bearing on revenue.

In p2pool, all valid shares must pay out to the finders of the last 8640 shares according to how much hashpower each miner contributed.

Each attempt at finding a share or block is statistically independent of all other attempts. It's like rolling a 1 quadrillion sided die and hoping that it rolls a 1. Your chance to roll a 1 is independent of how many times you've rolled before. If someone else rolls a 1, you have no idea how many die rolls they actually took to get that 1, but you know that it would take 1 quadrillion rolls on average, and you know they did at least 1 roll. Same with finding a share. We don't know or care how many unsuccessful attempts there were; we only care about the successful attempts.

Your miner accepts a block template (or share template) from p2pool, then does the following nested loop repeatedly:

While (is_mining):
    for i in rage(2**32-1):
        header.nonce += 1
        if sha256(sha256(header)) < target:
             submit(header)
    block.transactions[0].extranonce1 += 1
    header.merkleroot = compute_merkle_root(block.transactions)

**Comment by ikolubr** [comment]
@jtoomim Could you point me in some direction on how the tx works on p2pool? As I said, I'm trying to understand the whole thing... I'm trying to make p2pool work with VERGE... I got a few things changed on the p2pool coda as the VERGE wallet report few things differently... like version it is output as protocolversion and few other minor changes. I got stuck on this error, that I think it is related how the tx is dealt, but I have no idea where to check or how to check or even where to start..... if you could point me somewhere that would be great.

Thank you,

**Pull request closed without merging**